android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/translate/GeckoTranslationState.kt

```diff
@@ -0,0 +1,21 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.concept.engine.translate
+
+/**
+* The representation of the Gecko translations engine toolkit state.
+*
+* @property detectedLanguages Detected information about preferences and page information.
+* @property error If an error state occurred or an error was reported by toolkit.
+* @property isEngineReady If the toolkit translation engine is primed for use or will need to be loaded.
+* @property requestedTranslationPair The language pair to translate. Usually populated after first request.
+*/
+
+data class GeckoTranslationState(
```

Comment:
nit: I think we have to omit the "Gecko" prefix, as concept data classes are engine agnostic, they are used to represent any data coming from any `Engine` implementation.

Reply:
Thanks! We have a bit of a namespace collision here with `TranslationState` coming from [GeckoView](https://searchfox.org/mozilla-central/rev/a9cb8b07d4cf4290e693bc31ba23412d0f042cb8/mobile/android/geckoview/src/main/java/org/mozilla/geckoview/TranslationsController.java#942) versus the browser's [translation state](https://searchfox.org/mozilla-mobile/source/firefox-android/android-components/components/browser/state/src/main/java/mozilla/components/browser/state/state/TranslationsState.kt). 

 I'm open to any naming, but we need to decide based on that information. Some of the [original discussion is here too](https://github.com/mozilla-mobile/firefox-android/pull/4231#discussion_r1380306894).

Reply:
Sorry, I'm not that familiar with the translation APIs do we have some technical docs where I can get more context about the flow of translations. From what you mentioned, it looks like we are could have two sources of request for translation, I would like to get more context about it 😅 ?

Reply:
Thanks for taking a look! I really appreciate the review, esp. where there is so much context!

A translation can occur either:
1. If Fenix requests a translation or 
2. If the pref is set to automatically translate 

When the translation engine in toolkit does any operation it will call `onTranslationStateChange` to show a translation lifecycle event occurred. It gets called fairly regularly, maybe a couple of times per page load.

So, on an automatically translate call page load:
* Before Page load - onTranslationStateChange - no translation pair, most items will be null
* Page Load - onTranslationStateChange - will have a docLangTag of the detected language
* It translated -  onTranslationStateChange - will have the pair state it auto translated to

When Fenix calls translate, will see the same events coming from `onTranslationStateChange`; however, we will also have our progress [state set](https://searchfox.org/mozilla-mobile/rev/1eb55a6c369e7f208ce7c24934104b0fa24ac0ce/firefox-android/android-components/components/browser/state/src/main/java/mozilla/components/browser/state/state/TranslationsState.kt#16-17). (Test with some [examples](https://searchfox.org/mozilla-central/rev/a9cb8b07d4cf4290e693bc31ba23412d0f042cb8/mobile/android/geckoview/src/androidTest/java/org/mozilla/geckoview/test/TranslationsTest.kt#150) of the lifecycle too.)

Also, this made me thing of something that might influence how we organize this, `onTranslationStateChange` should probably also flip [`isTranslated`](https://searchfox.org/mozilla-mobile/rev/1eb55a6c369e7f208ce7c24934104b0fa24ac0ce/firefox-android/android-components/components/browser/state/src/main/java/mozilla/components/browser/state/state/TranslationsState.kt#15) when we see that an auto translate event occurred (criteria would be that we have a translation pair set, therefore it was translated). Which, makes the lifecycle a less separate. 

Let me know what state organization you'd prefer!

Reply:
>I guess then we have two type of translations, one Automatically and one Requested by the user, right?

Yes, that is right!

Reply:
We have a similar case for Prompt quests, where we have different type of prompts that shared common behaviour but each one has its particular properties. We modeled it creating the [PromptRequest.kt](https://github.com/mozilla-mobile/firefox-android/blob/1942ff65ff06690d8ed892ca3d514b3a6a6562d0/android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/prompt/PromptRequest.kt#L28).

It has a `sealed class PromptRequest`  which is the base class for all the requests, and all other requests are children `data class`s, which has its own properties and meaning,  for example [Alert](https://github.com/mozilla-mobile/firefox-android/blob/1942ff65ff06690d8ed892ca3d514b3a6a6562d0/android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/prompt/PromptRequest.kt#L76C5-L76C21).

Maybe having `sealed class TranslationRequest`, `data class AppTranslationRequest` and `data class EngineSuggestionTranslationRequest` 😅  could be an alternative? Also tagging @gabrielluong to hear his opinion.

Reply:
Just want to add one point of clarification, [GeckoTranslationState](https://searchfox.org/mozilla-central/rev/a9cb8b07d4cf4290e693bc31ba23412d0f042cb8/mobile/android/geckoview/src/main/java/org/mozilla/geckoview/TranslationsController.java#942) isn't a translations request. It is more a packet of information on the Gecko toolkit translations engine state. So, for example, on page load, we will receive information that the page can be translated or not. `GeckoTranslationState` changes independently of a translations requests, but will update in response to a translation request.

Reply:
Filed [1865919](https://bugzilla.mozilla.org/show_bug.cgi?id=1865919) as a follow up to refactor into a sealed class!